src/libraries/Microsoft.PowerFx.Core/Entities/External/InternalTableMetadata.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PowerFx.Core.App;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Core.Entities
+{
+    internal class InternalTableMetadata : IExternalTableMetadata
+    {        
+        private readonly AggregateType _type;
+
+        public InternalTableMetadata(AggregateType type, string name, string displayName, bool isReadOnly, string parameterPkColumnName = "")
+        {
+            Contracts.AssertNonEmpty(name);
+            Contracts.AssertNonEmpty(displayName);
+
+            _type = type;
+
+            Name = name;
+            DisplayName = displayName;
+            IsReadOnly = isReadOnly;
+                        
+            ParameterPkColumnName = parameterPkColumnName;
+        }        
+
+        public string Name { get; }
+
+        public string DisplayName { get; }
+
+        public string ParameterPkColumnName { get; }
+
+        public bool IsReadOnly { get; }
+
+        public IReadOnlyList<string> KeyColumns => _type._type.DisplayNameProvider.LogicalToDisplayPairs.Select(pair => pair.Key.Value).Where(col => col != "57dfb1b5-7d79-4046-a4da-fd831d5befe1-KeyId").ToList();
+
+        public IReadOnlyList<ColumnMetadata> Columns { get; }
+
+        public ColumnMetadata this[string columnName]
+        {
+            get
+            {
+                Contracts.AssertNonEmpty(columnName);
+                return TryGetColumn(columnName, out ColumnMetadata columnMetadata) ? columnMetadata : null;
+            }
+        }
+
+        public bool TryGetColumn(string columnName, out ColumnMetadata column)
+        {
+            string GetDisplayName(string fieldName) => _type._type.DisplayNameProvider == null || !_type._type.DisplayNameProvider.TryGetDisplayName(new DName(fieldName), out DName displayName) ? fieldName : displayName.Value;
+            DataFormat? ToDataFormat(DType dType) => dType.Kind switch
+            {
+                DKind.Record or DKind.Table or DKind.OptionSetValue => DataFormat.Lookup,
+                DKind.String or DKind.Decimal or DKind.Number or DKind.Currency => DataFormat.AllowedValues,
+                _ => null
+            };            
+
+            Contracts.AssertNonEmpty(columnName);
+
+            if (_type.TryGetFieldType(columnName, out FormulaType ft))
```

Comment:
>TryGetFieldType [](http://example.com/codeflow?start=22&length=15)

Should we call GetColumnCapability ? 

Ie  - how would a dervied type set the IsReadOnly flag?

Reply:
We can't do everything in just one PR 
This will be managed later, like the primary key or "required" (still unclear what this is)